src/test/java/com/neo4j/docker/TestUpgrade.java

```diff
@@ -4,60 +4,208 @@
+    }
+
+    @Test
+    void canUpgradeFromBeforeFilePermissionFix35() throws Exception
+    {
+        Neo4jVersion beforeFix = new Neo4jVersion( 3, 5, 3 );
+        String beforeFixImage = (TestSettings.EDITION == TestSettings.Edition.ENTERPRISE) ? "neo4j:3.5.3-enterprise" : "neo4j:3.5.3";
+        Assumptions.assumeTrue( TestSettings.NEO4J_VERSION.isNewerThan( beforeFix ), "test only applicable to latest 3.5 docker" );
+        Assumptions.assumeFalse( TestSettings.NEO4J_VERSION.isAtLeastVersion( Neo4jVersion.NEO4J_VERSION_400 ),
+                                 "test only applicable to latest 3.5 docker" );
+
+        Path dataMount = HostFileSystemOperations.createTempFolder( "data-upgrade-" );
+        log.info( "created folder " + dataMount.toString() + " to test upgrade" );
+
+        try ( GenericContainer container = makeContainer( beforeFixImage ) )
+        {
+            HostFileSystemOperations.mountHostFolderAsVolume( container, dataMount, "/data" );
+            container.start();
+            DatabaseIO db = new DatabaseIO( container );
+            db.putInitialDataIntoContainer( user, password );
+        }
+
+        try ( GenericContainer container = makeContainer( TestSettings.IMAGE_ID ) )
+        {
+            HostFileSystemOperations.mountHostFolderAsVolume( container, dataMount, "/data" );
+            container.start();
+            DatabaseIO db = new DatabaseIO( container );
+            db.verifyDataInContainer( user, password );
+        }
+    }
+
+    @Test
+    void canUpgradeFromReleasedVersion() throws Exception
+    {
+        var targetNeo4jVersion = TestSettings.NEO4J_VERSION;
+
+        // If this is the very first in a new major series (i.e. a .0.0 release) then this test isn't expected to work
+        Assumptions.assumeFalse( targetNeo4jVersion.major == 0 && targetNeo4jVersion.minor == 0 );
+
+        // TODO: update this when moving to the next minor release
+        // I am taking a guess here that we will have published 4.1.0 by 1st of July
+        if ( Instant.now().isBefore( Instant.parse( "2020-07-01T00:00:00.00Z" ) ) )
+        {
+            // This sort-of-hack is necessary when we cut a new minor branch before the previous minor branch has been published to dockerhub.
+            Assumptions.assumeTrue( Neo4jVersion.NEO4J_VERSION_420.isNewerThan( targetNeo4jVersion ) );
+        }
+        String fromImageName = dockerImageToUpgradeFrom( targetNeo4jVersion );
+
+        var testMountDirPrefix = String.format( "upgrade-from-%s", fromImageName ).replaceAll( ":", "_" );
+        Path testMountDir = HostFileSystemOperations.createTempFolder( testMountDirPrefix );
+
+        // TODO: test /plugins and /ssl directories
+        Map<String,@NotNull Path> readonlyMounts = createDirectories( testMountDir, List.of( "/conf" ) );
+        Map<String,@NotNull Path> writableMounts = createDirectories( testMountDir, List.of( "/data", "/logs", "/metrics" ) );
```

Comment:
I've made a PR for some code refactoring I did but forgot to merge: #263
the logic you have here around `readonlyMounts` (which never gets used as a list), `writableMounts` and `allMounts` can be replaced by doing something like:
  
```
Path testMountDir = HostFileSystemOperations.createTempFolder( testMountDirPrefix );

HostFileSystemOperations.createTempFolderAndMountAsVolume( container, testMountDir, "conf", "/conf");
HostFileSystemOperations.createTempFolderAndMountAsVolume( container, testMountDir, "data", "/data");
HostFileSystemOperations.createTempFolderAndMountAsVolume( container, testMountDir, "logs", "/logs");
HostFileSystemOperations.createTempFolderAndMountAsVolume( container, testMountDir, "metrics", "/metrics");
```

plus I guess whatever's necessary to put it into a map of mount points to host paths.